### ISSUE TITLE
"Copy rich link" doesn't work on Google Docs

### ISSUE THREAD

**Issue by ianb** [mozilla/firefox-voice#1205]
If I execute "copy rich link" on a Google Docs page then nothing is being copied to the clipboard (whatever was previously on the clipboard stays there). Other sites work.

**Comment by shreyaa-s** [comment]
Hey @ianb can i get assigned to this one next?

**Comment by ianb** [comment]
@shreyaa-s Yep! My guess is that this and #1117 are related, and that it has to do with this code not working: https://github.com/mozilla/firefox-voice/blob/master/extension/intents/clipboard/contentScript.js#L28-L45

**Comment by shreyaa-s** [comment]
Hey @ianb !
I looked into the issue and the problem is not with the code but with google docs itself. The conventional methods of text selection do not work on Google Docs and that is why "Copy Rich Link" is not working on Google Docs. One way to make it work would be to import the GoogleDocsUtil library or at least a function which gives us the google doc.

**Comment by ianb** [comment]
@shreyaa-s we're making our own selection in this case, not selecting something in the document. But arguably what should happen is that we should pass the information back to `intents/clipboard/clipboard.js` and (in the background page) it should actually execute the copy command. Then it won't be running in someone else's page that may interfere with it.

**Comment by shreyaa-s** [comment]
Hey @ianb 
I'm stuck on this one. I've done extensive research for alternate ways to copy, the structure of google docs and google translate pages (in reference to #1117 ), gone through the code several times but I can't seem to figure out why it doesn't work. I've tried using googleDocs API, went through window and document variables and tried the clipboard API on mozilla firefox but still no luck. Can you help me out?

**Comment by ianb** [comment]
@shreyaa-s I was thinking you could move `copyElement` out of `clipboard/contentScript.js` and into `clipboard/clipboard.js`

So for instance you might change it to be:

```js
  types.copyRichLink = function() {
    const m = meta();
    return {
      type: "richLink",
      href: m.canonical,
      content: m.title,
    };
};
```

And then in clipboard.js:

```js
async function copy(context, copyType, complete = false) {
  const activeTab = await browserUtil.activeTab();
  await content.lazyInject(activeTab.id, [
    "/background/pageMetadata-contentScript.js",
    "/intents/saving/screenshotContentScript.js",
    "/intents/clipboard/contentScript.js",
  ]);
  if (complete) {
    await browserUtil.waitForDocumentComplete(activeTab.id);
  }
  const toCopy = await browser.tabs.sendMessage(activeTab.id, { type: "copy", copyType });
  if (toCopy && toCopy.type === "richLink") {
    copyRichLink(toCopy.href, toCopy.content);
  }
}

function copyRichLink(url, content) {
  const a = document.createElement("a"); 
  // ... etc
}
```

**Comment by shreyaa-s** [comment]
What does this do?
Does this stops the further execution of script till the copy link command is finished?
```
  if (complete) {
    await browserUtil.waitForDocumentComplete(activeTab.id);
  }
```

**Comment by shreyaa-s** [comment]
When is logged `toCopy`, it returned undefined.
I guess I should check `copyType` instead of `toCopy.type` for rickLink

**Comment by shreyaa-s** [comment]
Where does this function return value to?
Or how can i access the value returned by this function?
```javascript
  types.copyRichLink = function() {
    const m = meta();
    return {
      type: "richLink",
      href: m.canonical,
      content: m.title,
    };
};
```

**Comment by shreyaa-s** [comment]
hey @ianb i'm still stuck on this one. give it a read when you have a moment. Thanks!
